docs/current/guides/757394-use-services.md

```diff
@@ -0,0 +1,347 @@
+</TabItem>
+<TabItem value="Node.js">
+
+<Embed id="" />
+
+</TabItem>
+</Tabs>
+
+When a service is bound to a container, it also conveys to any outputs of that container, such as files or directories. The service will be started whenever the output is used, so you can also do things like this:
+
+<Tabs groupId="language" className="embeds">
+<TabItem value="Go">
+
+<Embed id="Qoljk5SEPuu" />
+
+</TabItem>
+
+<TabItem value="Python">
+
+<Embed id="" />
+
+</TabItem>
+<TabItem value="Node.js">
+
+<Embed id="" />
+
+</TabItem>
+</Tabs>
+
+## Understanding the service lifecycle
+
+If you're not interested in what's happening in the background, you can skip this section and just trust that services are running when they need to be. If you're interested in the theory, keep reading.
+
+Consider this example:
+
+<Tabs groupId="language" className="embeds">
+<TabItem value="Go">
+
+<Embed id="JwuCvswjsEM" />
+
+</TabItem>
+
+<TabItem value="Python">
+
+<Embed id="" />
+
+</TabItem>
+<TabItem value="Node.js">
+
+<Embed id="" />
+
+</TabItem>
+</Tabs>
+
+Here's what happens on the last line:
+
+1. The client requests the `ping` container's stdout, which requires the container to run.
+1. Dagger sees that the `ping` container has a service binding, `redisSrv`.
+1. Dagger starts the `redisSrv` container, which recurses into this same process.
+1. Dagger waits for health checks to pass against `redisSrv`.
+1. Dagger runs the `ping` container with the `redis-srv` alias magically added to `/etc/hosts`.
+
+:::note
+Dagger cancels each service run after a 10 second grace period to avoid frequent restarts.
+:::
+
+It's worth noting that services are just containers, and all containers in Dagger have run-exactly-once semantics. Concurrent runs of the same container synchronize and attach to the same run with progress/logs multiplexed to each caller. Canceling a run only interrupts the container process when all runs are canceled, so new clients can come and go throughout a service run.
+
+Run-exactly-once semantics are very convenient. You don't have to come up with names and maintain instances of services; they're content-addressed, so you just use them by value. You also don't have to manage the state of the service; you can just trust that it will be running when needed and stopped when not.
+
+:::tip
+If you need multiple instances of a service, just attach something unique to each one, such as an instance ID.
+:::
+
+Let's put all this together in a full client-server example of running commands against a Redis service:
+
+<Tabs groupId="language" className="embeds">
+<TabItem value="Go">
+
+<Embed id="d4SIgAW2Feo" />
+
+</TabItem>
+
+<TabItem value="Python">
+
+<Embed id="" />
+
+</TabItem>
+<TabItem value="Node.js">
+
+<Embed id="" />
+
+</TabItem>
+</Tabs>
+
+Note that this example relies on the 10-second grace period, which you should try to avoid. It would be better to chain both commands together, which ensures that the service stays running for both:
```

Comment:
@vito why avoid?

Reply:
Depending on the 10-second grace period is risky because you never know what might cause a 10 second delay between calls to Dagger. The machine could be under a ton of load, the network between the client and Dagger could be flaky (remote Dagger?) or someone could add an extra call in between that takes a variable amount of time.

Reply:
Added